File: Excel VM/Parser.Lexer.fs
module private Parser.Lexer
open System

type stringClassifier = string -> bool
module CommonClassifiers =
  // util
  let getOccurances pred str =
    String.collect (fun e -> if pred e then string e else "") str
     |> String.length
  let (>>||) classifierA classifierB e = classifierA e || classifierB e
  let (>>&&) classifierA classifierB e = classifierA e && classifierB e
  let inline negate f = f >> not
  let isPrefix (a:string) (b:string) =  // a is a prefix of b?
    a.Length <= b.Length && a = b.[..a.Length - 1]
  let isSuffix (a:string) (b:string) =  // a is a suffix of b?
    a.Length <= b.Length && a = b.[b.Length - a.Length..]
  // does str start with delim1, end with delim2, and not have the delimiters overlap characters?
  let isDelimitedString (delim1:string) (delim2:string) (str:string) =
    delim1.Length + delim2.Length <= str.Length && isPrefix delim1 str && isSuffix delim2 str
  let entireString = String.forall
  let firstCharOfString isClassified (str:string) = isClassified str.[0]
  let firstSuffixOfString isClassified (str:string) = entireString isClassified str.[1..]
  let firstCharIsXAndSuffixIsY isClassified pred str =
    firstCharOfString isClassified str && pred str.[1..]
  // common classifiers (char)
  let isNumeral = Char.IsDigit     // 0 to 9
  let isAlphabetic = Char.IsLetter // a to z, A to Z, accented and Greek characters
  let isUndersc = (=) '_'
  let isDot = (=) '.'
  let isHyphen = (=) '-'
  let isValidForVariables = isNumeral >>|| isAlphabetic >>|| isUndersc
  let isWhitespaceChar = (=) ' ' >>|| (=) '\n' >>|| (=) '\t' >>|| (=) '\r'
  let isSingleQuote = (=) '''
  let isDoubleQuote = (=) '"'
  // common classifiers (string)
  let isVariableSuffix = entireString isValidForVariables
  let isInteger =   // first character is digit, rest are valid for variables
    firstCharIsXAndSuffixIsY isNumeral isVariableSuffix
  let isVariable =  // all characters are valid for variables, first is not digit
    firstCharIsXAndSuffixIsY (isValidForVariables >>&& negate isNumeral) isVariableSuffix
  let isNumericSuffix = entireString (isDot >>|| isValidForVariables)
  let isNumeric =   // first character is digit, rest are valid for variables or are dot
    firstCharIsXAndSuffixIsY isNumeral isNumericSuffix
  let isFloat = isNumeric >>&& (negate isInteger)  // is numeric, but contains a dot
  // note that isFloatSuffix === isNumericSuffix
  let isWhitespace = entireString isWhitespaceChar
  let isAnything = entireString (fun _ -> true)
  let failedToClassify message _ = failwith message
// rules which use classifiers
open CommonClassifiers
type tokenizeRule =
  |StickRule of (string -> string -> bool)
  |PriorityStickRule of (string -> string -> bool)    // gets checked first
  |SymbolRule of string
let makeRule (c1:stringClassifier) (c2:stringClassifier) =
  StickRule (fun s1 s2 -> c1 s1 && c2 s2)
let makeSymbolRule (symbol:string) = SymbolRule symbol
let makeDelimiterRule (symbol1:string) (symbol2:string) =
  PriorityStickRule (fun s1 s2 ->
    isPrefix symbol1 s1 && not (isDelimitedString symbol1 symbol2 s1)
   )
let tokenize (ruleset:(tokenizeRule*bool) list) (txt:string list) =
  let ruleset =  // put a sentry at the end to catch match failures
    ruleset @ [makeRule (failedToClassify "could not classify") isAnything, false]
  let txtWithSymbols, stickRules, priorityStickRules =
    // go through the rule set
    List.fold (fun (accTxt, accRules, accPriorityRules) -> function
      // stick-rules get filtered into two new lists
      |StickRule r, b -> accTxt, (r, b)::accRules, accPriorityRules
      |PriorityStickRule r, b -> accTxt, accRules, (r, b)::accPriorityRules
      // symbol rules are dealt with immediately: scan for each symbol
      |SymbolRule s, _ ->
        let rec tryGetPrefix = function   // try to match a with some prefix of b and split b
          |[], ll -> Some ([], ll)
          |_, [] -> None
          |a::_, b::_ when a <> b -> None
          |a::tl1, b::tl2 ->
            match tryGetPrefix (tl1, tl2) with
            |None -> None
            |Some (p, s) -> Some (b::p, s)
        let ss = List.ofSeq s |> List.map string
        let rec groupSymbols ll =     // symbols will not overwrite previous symbols, even if they match due to the way the matching list is constructed (one char per index, meaning prefixes with previously formed symbols will never be matched)
          match tryGetPrefix (ss, ll), ll with
          |Some (p, s), _ -> String.concat "" p::groupSymbols s
          |None, [] -> []
          |None, hd::tl -> hd::groupSymbols tl
        (groupSymbols accTxt, accRules, accPriorityRules)
     ) (txt, [], []) ruleset
  // priority rules go first
  let stickRules = List.rev priorityStickRules @ List.rev stickRules
  // make one scanning pass of the whole list of tokens, considering all stick-rules
  List.fold (fun acc nextToken ->
    match acc with
    |[] -> [nextToken]
    |lastToken::tl ->
      // one by one, apply each stick-rule to find the first one that matches
      let _, shouldStick = Seq.find (fun (stick, _) -> stick lastToken nextToken) stickRules
      // the bool for each stick-rule determines if a matching pair should always or never be concatenated
      if shouldStick
       then lastToken + nextToken::tl
       else nextToken::acc
   ) [] txtWithSymbols
   |> List.rev
module SpecialCases =
  let detectEscapeSequenceDoubleQuote =
    StickRule (fun s1 (s2:string) -> isPrefix "\"" s1 && isSuffix "\\\"" s1)
let createSingleLineComment delimiter = [
  makeSymbolRule delimiter, true
  makeDelimiterRule delimiter "\n", true
  makeRule (isDelimitedString delimiter "\n") isAnything, false
 ]
let createDelimitedComment delim1 delim2 = [
  makeSymbolRule delim1, true
  makeSymbolRule delim2, true
  makeDelimiterRule delim1 delim2, true
  makeRule (isDelimitedString delim1 delim2) isAnything, false
 ]
let createSymbol symbol = [
  makeSymbolRule symbol, true
  makeRule ((=) symbol) isAnything, false
  makeRule (fun e -> not (isPrefix e symbol)) (fun e -> isPrefix e symbol), false
 ]
// don't need to check for \" due to the context in which this is used
let createStrings = [
  SpecialCases.detectEscapeSequenceDoubleQuote, true
  makeDelimiterRule "\"" "\"", true
  makeRule (isDelimitedString "\"" "\"") isAnything, false
 ]
let createVariablesAndNumbers = [
  makeRule isNumeric isNumericSuffix, true
  makeRule isVariable isVariableSuffix, true
  makeRule isAnything isAnything, false
 ]

File: docs/scripts/Excel VM/Parser.Lexer.js
define(["exports", "fable-core/umd/Seq", "fable-core/umd/Symbol", "fable-core/umd/List", "fable-core/umd/Util", "fable-core/umd/String"], function (exports, _Seq, _Symbol2, _List, _Util, _String) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.createVariablesAndNumbers = exports.createStrings = exports.SpecialCases = exports.tokenizeRule = exports.CommonClassifiers = undefined;
  exports.makeRule = makeRule;
  exports.makeSymbolRule = makeSymbolRule;
  exports.makeDelimiterRule = makeDelimiterRule;
  exports.tokenize = tokenize;
  exports.createSingleLineComment = createSingleLineComment;
  exports.createDelimitedComment = createDelimitedComment;
  exports.createSymbol = createSymbol;

  var _Symbol3 = _interopRequireDefault(_Symbol2);

  var _List2 = _interopRequireDefault(_List);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  var CommonClassifiers = exports.CommonClassifiers = function (__exports) {
    var getOccurances = __exports.getOccurances = function (pred, str) {
      return Array.from((0, _Seq.collect)(function (e) {
        return pred(e) ? e : "";
      }, str)).join('').length;
    };

    var op_GreaterGreaterBarBar = __exports.op_GreaterGreaterBarBar = function (classifierA, classifierB, e) {
      if (classifierA(e)) {
        return true;
      } else {
        return classifierB(e);
      }
    };

    var op_GreaterGreaterAmpAmp = __exports.op_GreaterGreaterAmpAmp = function (classifierA, classifierB, e) {
      if (classifierA(e)) {
        return classifierB(e);
      } else {
        return false;
      }
    };

    var isPrefix = __exports.isPrefix = function (a, b) {
      if (a.length <= b.length) {
        return a === b.slice(null, a.length - 1 + 1);
      } else {
        return false;
      }
    };

    var isSuffix = __exports.isSuffix = function (a, b) {
      if (a.length <= b.length) {
        return a === b.slice(b.length - a.length, b.length);
      } else {
        return false;
      }
    };

    var isDelimitedString = __exports.isDelimitedString = function (delim1, delim2, str) {
      if (delim1.length + delim2.length <= str.length ? isPrefix(delim1, str) : false) {
        return isSuffix(delim2, str);
      } else {
        return false;
      }
    };

    var entireString = __exports.entireString = function (predicate) {
      return function (str) {
        return (0, _Seq.forAll)(predicate, str);
      };
    };

    var firstCharOfString = __exports.firstCharOfString = function (isClassified, str) {
      return isClassified(str[0]);
    };

    var firstSuffixOfString = __exports.firstSuffixOfString = function (isClassified, str) {
      return entireString(isClassified)(str.slice(1, str.length));
    };

    var firstCharIsXAndSuffixIsY = __exports.firstCharIsXAndSuffixIsY = function (isClassified, pred, str) {
      if (firstCharOfString(isClassified, str)) {
        return pred(str.slice(1, str.length));
      } else {
        return false;
      }
    };

    var isNumeral = __exports.isNumeral = function (arg00) {
      return !isNaN(parseInt(arg00));
    };

    var isAlphabetic = __exports.isAlphabetic = function (arg00) {
      return (/^[a-zA-Z]$/.test(arg00)
      );
    };

    var isUndersc = __exports.isUndersc = function () {
      var x = "_";
      return function (y) {
        return x === y;
      };
    }();

    var isDot = __exports.isDot = function () {
      var x = ".";
      return function (y) {
        return x === y;
      };
    }();

    var isHyphen = __exports.isHyphen = function () {
      var x = "-";
      return function (y) {
        return x === y;
      };
    }();

    var isValidForVariables = __exports.isValidForVariables = function () {
      var classifierA = function classifierA(e) {
        return op_GreaterGreaterBarBar(function (arg00) {
          return isNumeral(arg00);
        }, function (arg00_1) {
          return isAlphabetic(arg00_1);
        }, e);
      };

      return function (e_1) {
        return op_GreaterGreaterBarBar(classifierA, isUndersc, e_1);
      };
    }();

    var isWhitespaceChar = __exports.isWhitespaceChar = function () {
      var classifierA_2 = void 0;
      var classifierA_1 = void 0;
      var classifierA = void 0;
      var x = " ";

      classifierA = function classifierA(y) {
        return x === y;
      };

      var classifierB = void 0;
      var x_1 = "\n";

      classifierB = function classifierB(y_1) {
        return x_1 === y_1;
      };

      classifierA_1 = function classifierA_1(e) {
        return op_GreaterGreaterBarBar(classifierA, classifierB, e);
      };

      var classifierB_1 = void 0;
      var x_2 = "\t";

      classifierB_1 = function classifierB_1(y_2) {
        return x_2 === y_2;
      };

      classifierA_2 = function classifierA_2(e_1) {
        return op_GreaterGreaterBarBar(classifierA_1, classifierB_1, e_1);
      };

      var classifierB_2 = void 0;
      var x_3 = "\r";

      classifierB_2 = function classifierB_2(y_3) {
        return x_3 === y_3;
      };

      return function (e_2) {
        return op_GreaterGreaterBarBar(classifierA_2, classifierB_2, e_2);
      };
    }();

    var isSingleQuote = __exports.isSingleQuote = function () {
      var x = "'";
      return function (y) {
        return x === y;
      };
    }();

    var isDoubleQuote = __exports.isDoubleQuote = function () {
      var x = "\"";
      return function (y) {
        return x === y;
      };
    }();

    var isVariableSuffix = __exports.isVariableSuffix = entireString(isValidForVariables);

    var isInteger = __exports.isInteger = function (str) {
      return firstCharIsXAndSuffixIsY(function (arg00) {
        return isNumeral(arg00);
      }, isVariableSuffix, str);
    };

    var isVariable = __exports.isVariable = function () {
      var isClassified = void 0;

      var classifierB = function classifierB($var13) {
        return function (value) {
          return !value;
        }(function (arg00) {
          return isNumeral(arg00);
        }($var13));
      };

      isClassified = function isClassified(e) {
        return op_GreaterGreaterAmpAmp(isValidForVariables, classifierB, e);
      };

      return function (str) {
        return firstCharIsXAndSuffixIsY(isClassified, isVariableSuffix, str);
      };
    }();

    var isNumericSuffix = __exports.isNumericSuffix = entireString(function (e) {
      return op_GreaterGreaterBarBar(isDot, isValidForVariables, e);
    });

    var isNumeric = __exports.isNumeric = function (str) {
      return firstCharIsXAndSuffixIsY(function (arg00) {
        return isNumeral(arg00);
      }, isNumericSuffix, str);
    };

    var isFloat = __exports.isFloat = function () {
      var classifierB = function classifierB($var14) {
        return function (value) {
          return !value;
        }(isInteger($var14));
      };

      return function (e) {
        return op_GreaterGreaterAmpAmp(isNumeric, classifierB, e);
      };
    }();

    var isWhitespace = __exports.isWhitespace = entireString(isWhitespaceChar);
    var isAnything = __exports.isAnything = entireString(function (_arg1) {
      return true;
    });

    var failedToClassify = __exports.failedToClassify = function (message, _arg1) {
      throw new Error(message);
    };

    return __exports;
  }({});

  var tokenizeRule = exports.tokenizeRule = function () {
    function tokenizeRule(caseName, fields) {
      _classCallCheck(this, tokenizeRule);

      this.Case = caseName;
      this.Fields = fields;
    }

    _createClass(tokenizeRule, [{
      key: _Symbol3.default.reflection,
      value: function () {
        return {
          type: "Parser.Lexer.tokenizeRule",
          interfaces: ["FSharpUnion"],
          cases: {
            PriorityStickRule: ["function"],
            StickRule: ["function"],
            SymbolRule: ["string"]
          }
        };
      }
    }]);

    return tokenizeRule;
  }();

  (0, _Symbol2.setType)("Parser.Lexer.tokenizeRule", tokenizeRule);

  function makeRule(c1, c2) {
    return new tokenizeRule("StickRule", [function (s1) {
      return function (s2) {
        return c1(s1) ? c2(s2) : false;
      };
    }]);
  }

  function makeSymbolRule(symbol) {
    return new tokenizeRule("SymbolRule", [symbol]);
  }

  function makeDelimiterRule(symbol1, symbol2) {
    return new tokenizeRule("PriorityStickRule", [function (s1) {
      return function (s2) {
        return CommonClassifiers.isPrefix(symbol1, s1) ? !CommonClassifiers.isDelimitedString(symbol1, symbol2, s1) : false;
      };
    }]);
  }

  function tokenize(ruleset, txt) {
    var ruleset_1 = (0, _List.append)(ruleset, (0, _List.ofArray)([[makeRule(function () {
      var message = "could not classify";
      return function (arg10_) {
        return CommonClassifiers.failedToClassify(message, arg10_);
      };
    }(), CommonClassifiers.isAnything), false]]));
    var patternInput = (0, _Seq.fold)(function (tupledArg, _arg1) {
      if (_arg1[0].Case === "PriorityStickRule") {
        var r = _arg1[0].Fields[0];
        return [tupledArg[0], tupledArg[1], new _List2.default([r, _arg1[1]], tupledArg[2])];
      } else if (_arg1[0].Case === "SymbolRule") {
        var s = _arg1[0].Fields[0];

        var tryGetPrefix = function tryGetPrefix(_arg2) {
          if (_arg2[0].tail != null) {
            if (_arg2[1].tail != null) {
              if (function () {
                var b = _arg2[1].head;
                var a = _arg2[0].head;
                return !(0, _Util.equals)(a, b);
              }()) {
                return null;
              } else {
                var $var15 = _arg2[0].tail != null ? _arg2[1].tail != null ? [0, _arg2[0].head, _arg2[1].head, _arg2[0].tail, _arg2[1].tail] : [1] : [1];

                switch ($var15[0]) {
                  case 0:
                    var matchValue = tryGetPrefix([$var15[3], $var15[4]]);

                    if (matchValue != null) {
                      var s_1 = matchValue[1];
                      var p = matchValue[0];
                      return [new _List2.default($var15[2], p), s_1];
                    } else {
                      return null;
                    }

                  case 1:
                    throw new Error("C:\\Users\\Thefak\\Documents\\Visual Studio 2013\\Projects\\Excel VM\\build-docs\\../Excel VM/Parser.Lexer.fs", 74, 31);
                }
              }
            } else {
              return null;
            }
          } else {
            return [new _List2.default(), _arg2[1]];
          }
        };

        var ss = (0, _List.map)(function (value) {
          return value;
        }, (0, _Seq.toList)(s));

        var groupSymbols = function groupSymbols(ll) {
          var matchValue_1 = [tryGetPrefix([ss, ll]), ll];

          if (matchValue_1[0] == null) {
            if (matchValue_1[1].tail != null) {
              var tl = matchValue_1[1].tail;
              var hd = matchValue_1[1].head;
              return new _List2.default(hd, groupSymbols(tl));
            } else {
              return new _List2.default();
            }
          } else {
            var s_2 = matchValue_1[0][1];
            var p_1 = matchValue_1[0][0];
            return new _List2.default((0, _String.join)("", p_1), groupSymbols(s_2));
          }
        };

        return [groupSymbols(tupledArg[0]), tupledArg[1], tupledArg[2]];
      } else {
        var r_1 = _arg1[0].Fields[0];
        return [tupledArg[0], new _List2.default([r_1, _arg1[1]], tupledArg[1]), tupledArg[2]];
      }
    }, [txt, new _List2.default(), new _List2.default()], ruleset_1);
    var stickRules = (0, _List.append)((0, _List.reverse)(patternInput[2]), (0, _List.reverse)(patternInput[1]));
    return (0, _List.reverse)((0, _Seq.fold)(function (acc, nextToken) {
      if (acc.tail != null) {
        var patternInput_1 = (0, _Seq.find)(function (tupledArg_1) {
          return tupledArg_1[0](acc.head)(nextToken);
        }, stickRules);

        if (patternInput_1[1]) {
          return new _List2.default(acc.head + nextToken, acc.tail);
        } else {
          return new _List2.default(nextToken, acc);
        }
      } else {
        return (0, _List.ofArray)([nextToken]);
      }
    }, new _List2.default(), patternInput[0]));
  }

  var SpecialCases = exports.SpecialCases = function (__exports) {
    var detectEscapeSequenceDoubleQuote = __exports.detectEscapeSequenceDoubleQuote = new tokenizeRule("StickRule", [function (s1) {
      return function (s2) {
        return CommonClassifiers.isPrefix("\"", s1) ? CommonClassifiers.isSuffix("\\\"", s1) : false;
      };
    }]);
    return __exports;
  }({});

  function createSingleLineComment(delimiter) {
    return (0, _List.ofArray)([[makeSymbolRule(delimiter), true], [makeDelimiterRule(delimiter, "\n"), true], [makeRule(function () {
      var delim2 = "\n";
      return function (str) {
        return CommonClassifiers.isDelimitedString(delimiter, delim2, str);
      };
    }(), CommonClassifiers.isAnything), false]]);
  }

  function createDelimitedComment(delim1, delim2) {
    return (0, _List.ofArray)([[makeSymbolRule(delim1), true], [makeSymbolRule(delim2), true], [makeDelimiterRule(delim1, delim2), true], [makeRule(function (str) {
      return CommonClassifiers.isDelimitedString(delim1, delim2, str);
    }, CommonClassifiers.isAnything), false]]);
  }

  function createSymbol(symbol) {
    return (0, _List.ofArray)([[makeSymbolRule(symbol), true], [makeRule(function (y) {
      return symbol === y;
    }, CommonClassifiers.isAnything), false], [makeRule(function (e) {
      return !CommonClassifiers.isPrefix(e, symbol);
    }, function (e_1) {
      return CommonClassifiers.isPrefix(e_1, symbol);
    }), false]]);
  }

  var createStrings = exports.createStrings = (0, _List.ofArray)([[SpecialCases.detectEscapeSequenceDoubleQuote, true], [makeDelimiterRule("\"", "\""), true], [makeRule(function () {
    var delim1 = "\"";
    var delim2 = "\"";
    return function (str) {
      return CommonClassifiers.isDelimitedString(delim1, delim2, str);
    };
  }(), CommonClassifiers.isAnything), false]]);
  var createVariablesAndNumbers = exports.createVariablesAndNumbers = (0, _List.ofArray)([[makeRule(CommonClassifiers.isNumeric, CommonClassifiers.isNumericSuffix), true], [makeRule(CommonClassifiers.isVariable, CommonClassifiers.isVariableSuffix), true], [makeRule(CommonClassifiers.isAnything, CommonClassifiers.isAnything), false]]);
});